### PR TITLE
udevadm-wait: add --changed option

### PR DESCRIPTION
**Pull request by jimharris**

### CODE REVIEW

File: man/udevadm.xml [hunk 1]
```diff
@@ -961,6 +961,16 @@
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><option>--changed</option></term>
+          <listitem>
+            <para>When specified, the command wait for devices being changed instead of created or
+            initialized. If this is specified, <option>--initialized=</option> will be ignored.</para>
+
+            <xi:include href="version-info.xml" xpointer="v251"/>
```
Comment (yuwata) [inline man/udevadm.xml]: `v251` -> `v256`

File: shell-completion/zsh/_udevadm [hunk 2]
```diff
@@ -121,6 +121,7 @@ _udevadm_wait(){
         '--timeout=[Maximum number of seconds to wait for the devices being created.]' \
         '--initialized=[Wait for devices being initialized by systemd-udevd.]:boolean:(yes no)' \
         '--removed[Wait for devices being removed.]' \
+        '--changed[Wait for devices being changed.]' \
```
Comment (yuwata) [inline shell-completion/zsh/_udevadm:124]: When you touch this. Please also update completion for bash.

File: man/udevadm.xml [hunk 3]
```diff
@@ -961,6 +961,16 @@
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><option>--changed</option></term>
+          <listitem>
+            <para>When specified, the command wait for devices being changed instead of created or
```
Comment (yuwata) [inline man/udevadm.xml]: "changed" here is ambiguous. And, from the code, this implies "removed", even if the device does not exist when the command is invoked.

What is the use case here??

**Review by yuwata — CHANGES_REQUESTED** [review]
Please clarify its usecase. Also, please add test cases in TEST-17-UDEV.

File: man/udevadm.xml (hunk 3 above)
Comment (jimharris) [reply man/udevadm.xml]: Thanks @yuwata for the review. I've pushed a new rev with suggested changes and more justification in the commit message.

Comment (jimharris) [reply man/udevadm.xml]: I just realized I missed changes to TEST-17-UDEV. Please wait to review again until I push another rev.

**Comment by jimharris** [comment]
> Please clarify its usecase. Also, please add test cases in TEST-17-UDEV.

Can you provide some guidance on how to extend TEST-17-UDEV for this? Should I modify existing test case or add a new one?

**Comment by poettering** [comment]
I am not sure this makes much sense. A device usually sees one "add" and one "remove" events, but any amount of "change" events. Moreover, we insist on the liberty that "udevadm trigger" can be called any time, resulting in change events. Thus, the proposed command here is pretty useless, since you basically have zero idea what event you are waiting for. 

Hence, please elaborate on the usecase. What are you trying to do here?

**Comment by jimharris** [comment]
> Hence, please elaborate on the usecase. What are you trying to do here?

I added this just now to the commit message:

```
--changed will trigger on a CHANGE or REMOVE event. Primary use cases are
storage related:

1) Block device size change - for example, wait on change to
/sys/device/virtual/block/loop0 which will trigger when losetup attaches
or detaches a file to the loop0 device.
2) PCI device changes - for example, wait on change to sriov_numvfs. Since
NVMe SSDs are PCI devices and can be hot-removed, it is useful to treat
device removal (REMOVE event) same as a CHANGE event.
```

**Review by yuwata — CHANGES_REQUESTED** [review]
OK. Though I am not sure it works in general, but the use cases make some sense to me.
But, please do not imply --removed in --changed. But, make it possible to be _combined_ with --removed.
- If only --changed is specified, wait for only change uevent,
- If --changed is specified with --removed, then wait for change or remove uevent.
- Optional, but if --changed is specified with --added (of course please add it if you agree), then wait for change or add uevent.

**Comment by yuwata** [comment]
> > Please clarify its usecase. Also, please add test cases in TEST-17-UDEV.
> 
> Can you provide some guidance on how to extend TEST-17-UDEV for this? Should I modify existing test case or add a new one?

We have several test code for TEST-17-UDEV at test/units/testsuite-17.xx.sh. I think, adding simple test cases in testsuite-17.10.sh is enough.

**Comment by poettering** [comment]
> --changed will trigger on a CHANGE or REMOVE event. Primary use cases are
> storage related:
> 
> 1) Block device size change - for example, wait on change to
> /sys/device/virtual/block/loop0 which will trigger when losetup attaches
> or detaches a file to the loop0 device.

Well, as I said, this won't work. Anything is free to fire of change events, so how would you recognize *your* change event, and distinguish it from any other somebody might have triggered for the device?  As I said earlier, we expect that triggering change events on arbitrary devices is always safe, i.e. if something issues "udevadm trigger" on any device at any time, then *no* *bad* *things* *should* *happen*. But with this patch of yours, if people use "udevadm wait --changed", then bad things *will* happen, because you might mistake the change events somebody else triggers on your device from those you yourself are expecting.

> 2) PCI device changes - for example, wait on change to sriov_numvfs. Since
> NVMe SSDs are PCI devices and can be hot-removed, it is useful to treat
> device removal (REMOVE event) same as a CHANGE event.

I don't grok this, but the same issue applies.

Sorry, I am not convinced we should add this, because it is impossible to use correctly.

**Comment by poettering** [comment]
Or in other words: please provide a usecase where this makes sense to use, even if something in the background keeps doing udevadm trigger on all devices all the time. If you cannot provide a usecase, let's not merge this.

**Comment by jimharris** [comment]
> Or in other words: please provide a usecase where this makes sense to use, even if something in the background keeps doing udevadm trigger on all devices all the time. If you cannot provide a usecase, let's not merge this.

Thanks for the feedback. I need to consider this more, so closing this merge request for now.

**Pull request closed without merging**